### PR TITLE
fix: make Vite work with custom frontend directory

### LINKED ISSUE #14046: Vite does not work with custom frontend directory
### Description of the bug

When using Vite is a Spring Boot project with a custom frontend folder configured in `vaadin-maven-plugin`, the dev server seems to start, but the java application keeps restarting infinitely.

Main problem seems to be that in `vite.generated.ts` there a `root` key whose value is hard-coded to 'frontend'.
This can be fixed in `vite.config.ts`

```
const customConfig: UserConfigFn = (env) => ({
  // Here you can add custom Vite parameters
  // https://vitejs.dev/config/
  root: 'src/main/frontend'
});
```

But then there are failure in themes compilation, because they are still using `frontend/generated` directory instead to be nested into  the custom frontend folder.

The workaround for this is to set also the `generatedTsFolder` setting in `vaadin-maven-plugin`, pointing to a subfolder of the custom generated frontend dir.

```xml
<generatedTsFolder>src/main/frontend/generated</generatedTsFolder>
```

However this does not make Vite work because there are other issues with custom frontend folder.
This is slightly related to #12880).
For example some hard-coded relative path that assumes files are in `frontend/generated` (for example `../../target/xyz`).

Additional note: vite.generated.ts has the following confiuration that may also be wrong with custom frontend directory

```
const allowedFrontendFolders = [
  // other settings
  path.resolve(frontendFolder, '../node_modules')
];
```

### Minimal reproducible example

Download an application from start.vaadin.com with Vite enabled and add the following configuration to `vaadin-maven-plugin`.

```xml
<configuration>
    <frontendDirectory>src/main/frontend</frontendDirectory>
</configuration>
```

### Expected behavior

Vite dev serve works as when using default `frontend` directory

### Actual behavior

1) Dev server does not serve frontend files and Java server continously restarts itself
2) If workarounds are applyed, compilation fails due to "file not found" errors because of wrong relative paths

### Versions:
    - Vaadin / Flow version: 23.1.1
    - Java version: 11
    - OS version: Ubuntu 21.10
    - Browser version (if applicable):
    - Application Server (if applicable):
    - IDE (if applicable):
    - Development mode
    - Project from start.vaadin.com

### LINKED ISSUE #14102: Assets from npm packages not correctly resolved with Vite
### Description of the bug

Using assets from NPM packages in custom theme seems not to work with Vite.
On the resulting CSS the resource `url` is used as is and is not prefixed by theme path.

For example, applying configuration from [documentation](https://vaadin.com/docs/latest/components/ds-resources/customization/custom-theme-configuration/#assets), the following directive works fine with webpack, but not with Vite

```css
.icon-snowflake {
  background-image: url('./fontawesome/icons/snowflake.svg');
}
```
With webpack the URL that can be seen in the browser is `/themes/myapp/fontawesome/icons/snowflake.svg`, whereas with Vite remains unchanged and the resource is not loaded.

The problem seems to be that with webpack there is a `theme-loader` plugin (https://github.com/vaadin/flow/blob/master/flow-server/src/main/resources/plugins/theme-loader/theme-loader.js) that performs some url rewriting, but Vite seems not to have such configuration.

### Minimal reproducible example

Unpack the attached application, run `mvn` and open the browser at http://localhost:8080
You should see CSS rule applied, but the page is blank. 
If switching to webpack by disabling Vite feature flag, the application works as expected

[my-app.zip](https://github.com/vaadin/flow/files/9009072/my-app.zip)

### Expected behavior

See big snowflakes on the screen :)
The resources from the npm packages are correctly resolved and can be downloaded by the browser.

### Actual behavior

The browser page is blank.
Resources from npm packages are not served (HTTP 404)

### Versions:
    - Vaadin / Flow version: 23.1
    - Java version: 11
    - Development mode
    - Project from start.vaadin.com

### PR DESCRIPTION
**Pull request by mcollovati**
## Description

Fixes Vite configuration to work with a custom frontend directory.
Also rewrite theme css to resolve urls for assets defined in theme.json
    
Fixes #14046
Fixes #14102

## Type of change

- [X] Bugfix
- [ ] Feature

## Checklist

- [X] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [X] I have added a description following the guideline.
- [X] The issue is created in the corresponding repository and I have referenced it.
- [X] I have added tests to ensure my change is effective and works as intended.
- [X] New and existing tests are passing locally with my change.
- [X] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Comment by mcollovati** [comment]
Based on #14144

File: flow-tests/test-custom-frontend-directory/test-themes-custom-frontend-directory/src/main/vite-resources/vaadin-featureflags.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+com.vaadin.experimental.viteForFrontendBuild=true
```
Comment (taefi) [inline flow-tests/test-custom-frontend-directory/test-themes-custom-frontend-directory/src/main/vite-resources/vaadin-featureflags.properties:1]:
```suggestion
com.vaadin.experimental.viteForFrontendBuild=true

```

File: flow-server/src/main/resources/plugins/theme-loader/theme-loader-utils.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// Collect groups [url(] ['|"]optional './|../', file part and end of url
+const urlMatcher = /(url\(\s*)(\'|\")?(\.\/|\.\.\/)(\S*)(\2\s*\))/g;
+
+
+function assetsContains(fileUrl, themeFolder, logger) {
+  const themeProperties = getThemeProperties(themeFolder);
+  if (!themeProperties) {
+    logger.debug('No theme properties found.');
+    logger.log('No theme properties found.');
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/resources/plugins/theme-loader/theme-loader-utils.js:12]: Good catch, thanks! I added `log` for local debug. It should be removed

File: flow-server/src/main/resources/plugins/theme-loader/theme-loader-utils.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    return false;
+  }
+  const assets = themeProperties['assets'];
+  if (!assets) {
+    logger.debug('No defined assets in theme properties');
+    logger.log('No defined assets in theme properties');
```
[the remark this answers is not in the document]
Comment (mcollovati) [reply flow-server/src/main/resources/plugins/theme-loader/theme-loader-utils.js:18]: Good catch, thanks! I added `log` for local debug. It should be removed

**Review by taefi — APPROVED** [review]

**Review by mshabarov — APPROVED** [review]

**Comment by mcollovati** [comment]
Picking the change to 23.0 requires some huge refactoring and it also needs other changes to be back-ported.
Since V23.0 will be out of support soon, I'll propose to not port this one to 23.0

**Pull request merged**